compiler/damlc/lib/DA/Cli/Damlc/Test.hs

```diff
@@ -154,89 +168,200 @@ printSummary color res =
+    report :: String -> (LocalOrExternal -> Bool) -> Report
+    report groupName pred =
+        let filteredResults :: [(LocalOrExternal, [(VirtualResource, Either SSC.Error SS.ScenarioResult)])]
+            filteredResults = filter (pred . fst) results
+            definedTemplatesInside = M.keysSet $ templatesDefinedIn $ map fst filteredResults
+            definedChoicesInside = M.keysSet $ choicesDefinedIn $ map fst filteredResults
+            exercisedInside = foldMap (exercisedChoices . snd) filteredResults
+            createdInside = foldMap (createdTemplates . snd) filteredResults
+            internalExercisedAnywhere = allExercisedChoices `S.intersection` definedChoicesInside
+            internalExercisedInternal = exercisedInside `S.intersection` definedChoicesInside
+            externalExercisedInternal = exercisedInside `S.difference` definedChoicesInside
+            internalCreatedAnywhere = allCreatedTemplates `S.intersection` definedTemplatesInside
+            internalCreatedInternal = createdInside `S.intersection` definedTemplatesInside
+            externalCreatedInternal = createdInside `S.difference` definedTemplatesInside
+        in
+        Report
+            { groupName
+            , definedChoicesInside
+            , internalExercisedAnywhere
+            , internalExercisedInternal
+            , externalExercisedInternal
+            , definedTemplatesInside
+            , internalCreatedAnywhere
+            , internalCreatedInternal
+            , externalCreatedInternal
+            }
+
+    printReport :: Report -> IO ()
+    printReport
+        Report
+            { groupName
+            , definedChoicesInside
+            , internalExercisedAnywhere
+            , definedTemplatesInside
+            , internalCreatedAnywhere
+            } =
+        let percentage i j
+              | j > 0 = show (round @Double $ 100.0 * (fromIntegral i / fromIntegral j) :: Int) <> "%"
+              | otherwise = "100%"
+            frac msg a b = msg ++ ": " ++ show a ++ " / " ++ show b
+            pct msg a b = frac msg a b ++ " (" ++ percentage a b ++ ")"
+            indent = ("  " ++)
+            header = groupName ++ ":"
+            body1 =
+                [ pct "choices" (S.size internalExercisedAnywhere) (S.size definedChoicesInside)
+                , pct "templates" (S.size internalCreatedAnywhere) (S.size definedTemplatesInside)
+                ]
+            body2
+              | not getShowCoverage = []
+              | otherwise =
+                [ "templates never created:" ] <>
+                map (indent . printTemplateIdentifier) (S.toList $ definedTemplatesInside `S.difference` internalCreatedAnywhere) <>
+                ["choices never executed:"] <>
+                map (indent . printChoiceIdentifier) (S.toList $ definedChoicesInside `S.difference` internalExercisedAnywhere)
+            msg = unlines $ header : map indent (body1 ++ body2)
+        in
+        putStrLn msg
```

Comment:
I think this would look even better using [`DA.Pretty`](https://github.com/digital-asset/daml/blob/export-exercise-interface/libs-haskell/da-hs-base/src/DA/Pretty.hs) but we can do that later